keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+other match related properties in the future).
+
+"matching schema template":
+
+```go
+// TODO: Add this as a struct into the Kubernetes codebase so it can easily be
+// imported?
+type Match struct {
+  Rules []admissionregisterationv1.RuleWithOperations `json:"rules,omitempty" protobuf:"bytes,1,rep,name=rules"`
+  NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty" protobuf:"bytes,2,opt,name=namespaceSelector"`
+  ObjectSelector *metav1.LabelSelector `json:"objectSelector,omitempty" protobuf:"bytes,3,opt,name=objectSelector"`
+}
+```
+
+Example usage:
+
+```go
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ReplicaLimit struct {
+  MaxReplicas int32 `json:"maxReplicas" protobuf:"varint,1,name=maxReplicas"`
+
+  Match Match `json:"match,omitempty" protobuf:"bytes,2,name=match"`
+}
+```
+
+For comparison (provided by @liggitt):
+
+- RBAC, priority & fairness, and admission webhook configurations all use
+ordered lists of rules, and first match allows/matches.
+- Audit configuration also uses an ordered list of rules, but can make a match
+exclude auditing... this approach could address the "ability to skip specific
+resource types" mentioned above.
+
+TODO: What exact matching mechanisms will be supported? Right now this KEP
+shows the use of the same match rules that admission webhooks use.
+
+TODO: Should policy definitions also have match criteria? Match criteria at the
+policy definition level could be used constrain what resources the policy may
+apply to across the entire cluster (like the above "kind constraint" but more
+generally). This would allow policy authors to constrain use of the policy to
+the resources it was designed for.
+
+TODO: Should "standalone policies" be supported? I.e. should it be possible to
+declare a policy definition that has no `spec.config` but has it's own match
+criteria? Right now a policy requires a minimum of three resources to do
+anything (policy definition, configuration CRD, at least one policy
+configuration).
+
+`MatchPolicy` will implicitly be `Equivalent`, and will not be configurable,
+but it will be possible to opt-out of the behavior by inspecting the
+GVK of the admitted resource in the CEL expression.
+
+xref:
+https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules
+
+#### Type safety
+
+To keep failure policy easy to reason about, and to continue to use CEL in a
+type-safe way we propose:
+
+- If a ValidatingAdmissionPolicy has a `spec.constraints` that matches a single
+  GVK, the CEL expression is allowed access to the full object in a typesafe
+  way. Otherwise, the CEL expression is allowed access to the metadata only.
+- If there are any type checking errors (or if the CRD for the matched GVK does
+  not exist):
+  - When a CRD is created/updated: The type check errors are detected by an
```

Comment:
What is the impact if a CRD is updated such that pre-existing policies are put in an error state? Are requests denied?

Reply:
I think that's the right thing to do.  If the CRD updated intended to make the policy more restrictive (but accidentally broke the policy) then denying requests is less risky than allowing them.  And if the API server were restarted, it can become impossible to know what the previous CRD state was. I'll clarify.

Reply:
One clarification here: All requests fail, but they are only denied if the `FailurePolicy` is set to `Fail`.